cmd/e2e-test/carbon_clickhouse.go

```diff
@@ -0,0 +1,124 @@
+package main
+
+import (
+	"fmt"
+	"io/ioutil"
+	"os"
+	"os/exec"
+	"path"
+	"path/filepath"
+	"strconv"
+	"syscall"
+	"text/template"
+
+	"github.com/phayes/freeport"
+)
+
+type CarbonClickhouse struct {
+	Binary    string `toml:"binary"`
+	ConfigTpl string `toml:"template"`
+
+	storeDir   string    `toml:"-"`
+	configFile string    `toml:"-"`
+	address    string    `toml:"-"`
+	cmd        *exec.Cmd `toml:"-"`
+}
+
+func (c *CarbonClickhouse) Start(clickhouseAddr string) error {
+	if c.cmd != nil {
+		return fmt.Errorf("carbon-clickhouse already started")
+	}
+
+	if len(c.Binary) == 0 {
+		c.Binary = "./carbon-clickhouse"
+	}
+	if len(c.ConfigTpl) == 0 {
+		return fmt.Errorf("carbon-clickhouse config template not set")
+	}
+
+	var err error
+
+	c.storeDir, err = ioutil.TempDir("", "carbon-clickhouse")
+	if err != nil {
+		return err
+	}
+	port, err := freeport.GetFreePort()
```

Comment:
I've taken a look at freeport lib. Personally, I would just token a function out of there with credits AND used the `127.0.0.1:0` as the address, since you use it later.

The usage of `localhost` during the probing and `127.0.0.1` as the address further is error-prone. You never know how people set up their own DNS

And to reduce the boilerplate, you can use it as `getLocalhostAdress`, which will return `listener.Addr().String()`